app/src/main/cpp/skyline/nce.cpp

```diff
@@ -12,405 +13,311 @@
-                    ThreadTrace();
-
-                    state.ctx->state = ThreadState::WaitRun;
-                    break;
+            auto function{kernel::svc::SvcTable[svc]};
+            if (function) {
+                state.logger->Debug("SVC called 0x{:X}", svc);
+                (*function)(state);
+            } else {
+                throw exception("Unimplemented SVC 0x{:X}", svc);
+            }
+        } catch (const signal::SignalException &e) {
+            if (e.signal != SIGINT) {
+                state.logger->Error("{} (SVC: 0x{:X})", e.what(), svc);
+                if (state.thread->id) {
+                    signal::BlockSignal({SIGINT});
+                    state.process->Kill(false);
                 }
             }
+            abi::__cxa_end_catch(); // We call this prior to the longjmp to cause the exception object to be destroyed
+            std::longjmp(state.thread->originalCtx, true);
         } catch (const std::exception &e) {
-            state.logger->Error(e.what());
-        } catch (...) {
-            state.logger->Error("An unknown exception has occurred");
-        }
-
-        if (!Halt) {
-            if (thread == state.process->pid) {
-                JniMtx.lock(GroupMutex::Group::Group2);
-
-                state.os->KillThread(thread);
-                Halt = true;
-
-                JniMtx.unlock();
-            } else {
-                state.os->KillThread(thread);
+            state.logger->Error("{} (SVC: 0x{:X})", e.what(), svc);
+            if (state.thread->id) {
+                signal::BlockSignal({SIGINT});
+                state.process->Kill(false);
             }
+            abi::__cxa_end_catch();
+            std::longjmp(state.thread->originalCtx, true);
         }
-
-        state.jvm->DetachThread();
     }
 
-    NCE::NCE(DeviceState &state) : state(state) {}
-
-    NCE::~NCE() {
-        for (auto &thread : threadMap)
-            thread.second->join();
-    }
-
-    void NCE::Execute() {
-        try {
-            while (true) {
-                std::lock_guard guard(JniMtx);
-                if (Halt)
-                    break;
-                state.gpu->Loop();
-            }
-        } catch (const std::exception &e) {
-            state.logger->Error(e.what());
-        } catch (...) {
-            state.logger->Error("An unknown exception has occurred");
-        }
+    void NCE::SignalHandler(int signal, siginfo *info, ucontext *context, void **tls) {
+        if (*tls) {
+            const auto &state{*reinterpret_cast<ThreadContext *>(*tls)->state};
+            if (signal != SIGINT) {
+                state.logger->Warn("Thread #{} has crashed due to signal: {}", state.thread->id, strsignal(signal));
+
+                std::string raw;
+                std::string trace;
+                std::string cpuContext;
+
+                const auto &ctx{reinterpret_cast<ucontext *>(context)->uc_mcontext};
```

Comment:
is there a reason for casting ucontext * to ucontext * here?

Reply:
No, it's a remnant from past code back when that argument was `void *` (That's what it should be due to `sa_action` in the `sigaction` structure but it is always `ucontext *` so I directly made it `ucontext *`)